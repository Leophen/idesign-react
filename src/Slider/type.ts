import { placementType } from '../Popup/type';

export interface SliderProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 滑块布局方向
   * @default horizontal
   */
  layout?: 'vertical' | 'horizontal';
  /**
   * 滑块固定值（受控）
   */
  value?: number | number[];
  /**
   * 滑块默认值（非受控）
   */
  defaultValue?: number | number[];
  /**
   * 是否禁用滑块
   * @default false
   */
  disabled?: boolean;
  /**
   * 滑块范围最大值
   * @default 100
   */
  max?: number;
  /**
   * 滑块范围最小值
   * @default 0
   */
  min?: number;
  /**
   * 步长
   * @default 1
   */
  step?: number;
  /**
   * 是否为范围滑块
   * @default false
   */
  range?: boolean;
  /**
   * 是否隐藏数值提示
   * @default false
   */
  hideTip?: boolean;
  /**
   * 数值提示出现位置
   * @default top
   */
  placement?: placementType;
  /**
   * 滑块值变化时触发
   */
  onChange?: (value: number) => void;
}

export interface SliderBtnProps extends SliderProps {
  /**
   * 当前值
   * @default 0
   */
  currentVal?: number;
  /**
   * 是否正选中滑条
   * @default false
   */
  downSlider?: boolean;
}
