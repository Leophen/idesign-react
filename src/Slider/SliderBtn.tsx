import React, { useEffect, useRef, useState } from 'react';
import './index.scss';
import Popup from '../Popup';
import { SliderBtnProps } from './type';

const SliderBtn: React.FC<SliderBtnProps> = (props) => {
  const {
    layout = 'horizontal',
    currentVal = 0,
    max = 100,
    min = 0,
    downSlider = false,
    hideTip = false,
    placement = 'top',
  } = props

  const [enterShowTip, setEnterShowTip] = useState(false)

  const btnRef = useRef<HTMLDivElement>(null)
  const [rect, setRect] = useState({
    width: 0,
    height: 0
  })
  useEffect(() => {
    rect.width = btnRef.current?.getBoundingClientRect().width || 0
    rect.height = btnRef.current?.getBoundingClientRect().height || 0
    setRect({ ...rect })
  }, [])

  const getBtnStyle = () => {
    if (layout === 'horizontal') {
      return { left: `calc(${(currentVal - min) / (max - min) * 100}% - ${rect.width / 2}px)` }
    } else {
      return { top: `calc(${(currentVal - min) / (max - min) * 100}% - ${rect.height / 2}px)` }
    }
  }

  return (
    <Popup
      content={currentVal}
      updateLocation={currentVal}
      trigger="hover"
      visible={!hideTip && (downSlider || enterShowTip)}
      placement={placement}
    >
      <div
        className="i-slider__button"
        ref={btnRef}
        style={getBtnStyle()}
        onMouseEnter={() => setEnterShowTip(true)}
        onMouseLeave={() => setEnterShowTip(false)}
      />
    </Popup>
  )
}

export default SliderBtn;
