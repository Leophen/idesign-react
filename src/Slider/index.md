---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 14
---

# Slider 滑块

用于通过滑动来选择合适数值的滑动组件。

## 基本用法

```tsx
import React, { useState } from 'react'
import { Slider } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState(30)

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div style={{ width: 600 }}>
      <h4>无默认值</h4>
      <Slider />
      <h4>有默认值</h4>
      <Slider defaultValue={value} />
      <h4>有受控值</h4>
      <Slider value={value} />
      <h4>一般用法</h4>
      <Slider
        value={value}
        onChange={handleChange}
      />
    </div>
  )
}

export default App
```

## 禁用状态

通过 `disabled` 属性控制滑块为禁用状态。

```tsx
import React from 'react'
import { Slider } from 'idesign-react'

const App = () => {
  return (
    <div style={{ width: 600 }}>
      <Slider defaultValue={60} disabled />
    </div>
  )
}

export default App
```

## 自定义边界

通过 `max` 和 `min` 属性控制滑块最大值和最小值。

```tsx
import React from 'react'
import { Slider } from 'idesign-react'

const App = () => {
  return (
    <div style={{ width: 600 }}>
      <h4>max 设为 10</h4>
      <Slider defaultValue={5} max={10} />
      <h4>min 设为 30</h4>
      <Slider defaultValue={60} min={30} />
      <h4>同时设置</h4>
      <Slider defaultValue={100} max={200} min={50} />
    </div>
  )
}

export default App
```

## 自定义步长

通过 `step` 属性控制滑块步长。

```tsx
import React from 'react'
import { Slider } from 'idesign-react'

const App = () => {
  return (
    <div style={{ width: 600 }}>
      <h4>step 设为 0.6</h4>
      <Slider defaultValue={60} step={0.6} />
      <h4>step 设为 3</h4>
      <Slider defaultValue={33} step={3} />
      <h4>step 设为 10</h4>
      <Slider defaultValue={50} step={10} />
    </div>
  )
}

export default App
```

## 纵向滑块

通过 `layout` 属性控制滑块方向，默认为 `horizontal`。

```tsx
import React, { useState } from 'react'
import { Slider } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState(30)

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <div className="idesign-demo-block-row" style={{ height: 200 }}>
      <h4>无默认值 →</h4>
      <Slider layout="vertical" />
      <h4>有默认值 →</h4>
      <Slider layout="vertical" defaultValue={value} />
      <h4>有受控值 →</h4>
      <Slider layout="vertical" value={value} />
      <h4>一般用法 →</h4>
      <Slider
        layout="vertical"
        value={value}
        onChange={handleChange}
      />
    </div>
  )
}

export default App
```

## 范围滑块

通过 `range` 属性指定为范围滑块，注意 `value` 的值需为数组，例如 `[0, 100]`。

```tsx
import React from 'react'
import { Slider } from 'idesign-react'

const App = () => {
  return (
    <div style={{ width: 600 }}>
      <Slider range defaultValue={[30, 90]} />
    </div>
  )
}

export default App
```

## 隐藏数值提示

通过 `hideTip` 属性隐藏数值提示。

```tsx
import React from 'react'
import { Slider } from 'idesign-react'

const App = () => {
  return (
    <div style={{ width: 600 }}>
      <Slider defaultValue={60} hideTip />
    </div>
  )
}

export default App
```

## 数值提示位置

通过 `placement` 属性自定义数值提示位置。

```tsx
import React from 'react'
import { Slider } from 'idesign-react'

const App = () => {
  return (
    <div className="idesign-demo-block-row" style={{ width: 600 }}>
      <Slider defaultValue={20} placement='left' />
      <Slider defaultValue={40} placement='top' />
      <Slider defaultValue={60} placement='right' />
      <Slider defaultValue={80} placement='bottom' />
    </div>
  )
}

export default App
```

<API />
