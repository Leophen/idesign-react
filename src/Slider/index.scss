.i-slider {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  width: 100%;
  height: 20px;
  cursor: pointer;
  &.i-slider__layout-vertical {
    width: 20px;
    height: 100%;
    .i-slider__bar {
      width: 6px;
      height: 100%;
    }
    .i-slider__button {
      top: 0;
      left: 50%;
      transform: translateX(-50%);
    }
  }
  &.i-slider__disabled {
    cursor: not-allowed;
    .i-slider__bar {
      background-color: var(--i-primary-disabled);
    }
    .i-slider__button {
      border-color: var(--i-primary-disabled);
    }
  }
}

.i-slider__bar {
  position: relative;
  width: 100%;
  height: 6px;
  overflow: hidden;
  background-color: var(--i-primary);
  border-radius: 3px;
}

.i-slider__bar-active {
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: var(--i-bg-active);
}

.i-slider__button {
  position: absolute;
  top: 50%;
  left: 0;
  box-sizing: border-box;
  width: 20px;
  height: 20px;
  background: #fff;
  border: 2px solid var(--i-primary);
  border-radius: 50%;
  transform: translateY(-50%);
}
