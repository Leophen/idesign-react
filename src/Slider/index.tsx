import React, { useEffect, useMemo, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import useDefault from '../hooks/useDefault';
import _ from 'lodash';
import SliderBtn from './SliderBtn';
import { SliderProps } from './type';

// 取数组中 最接近传入值的数
const getNearNum = (numArr: number[], num: number) => {
  const newArr = _.cloneDeep(numArr)
  newArr.push(num)
  newArr.sort()
  let nearNum = 0
  for (let i = 0, len = newArr.length; i < len; i++) {
    if (newArr[i] === num) {
      if (i === 0) {
        nearNum = numArr[0]
      } else if (i === len - 1) {
        nearNum = numArr[numArr.length - 1]
      } else {
        if (Math.abs(newArr[i - 1] - num) < Math.abs(newArr[i + 1] - num)) {
          nearNum = newArr[i - 1]
        } else {
          nearNum = newArr[i + 1]
        }
      }
    }
  }
  return nearNum
}

const Slider: React.FC<SliderProps> = (props) => {
  const {
    className,
    style,
    layout = 'horizontal',
    value,
    defaultValue = !props.range ? (props.min || 0) : ([props.min || 0, props.max || 0]),
    disabled = false,
    max = 100,
    min = 0,
    step = 1,
    range = false,
    hideTip = false,
    placement = 'top',
    onChange,
  } = props;

  const [innerValue, setInnerValue] = useDefault(value, defaultValue, onChange);

  const slider = useRef<HTMLDivElement>(null)
  const [rect, setRect] = useState({
    left: 0,
    top: 0,
    width: 0,
    height: 0
  })

  const [downStatus, setDownStatus] = useState(false)

  // 是否为移动状态
  const setMoving = (ifMoving: boolean) => {
    setDownStatus(ifMoving)
    if (ifMoving) {
      document.body.style.userSelect = 'none'
    } else {
      document.body.style.userSelect = ''
    }
  }

  // 精度
  const [stepPrecision, setStepPrecision] = useState(0)

  useEffect(() => {
    setStepPrecision(String(step).split('.')[1]?.length)
  }, [])

  // 所有可取值的位置
  const positionArr = useMemo(() => {
    const result = [];
    for (let i = min; i <= max; i = _.add(i, step)) {
      result.push((i - min) / (max - min));
    }
    return result;
  }, [max, min, step]);

  const positTurnVal = (movePercent: number) => {
    return min + _.round((max - min) * getNearNum(positionArr, movePercent), stepPrecision);
  }

  const lastVal = useRef(0)
  const handleSliderMove = _.throttle((e: any) => {
    let move = 0
    let minMove = 0
    let maxMove = 0

    if (layout === 'horizontal') {
      move = e.clientX - rect.left
      maxMove = rect.width
    } else {
      move = e.clientY - rect.top
      maxMove = rect.height
    }
    move < (minMove) && (move = minMove)
    move > (maxMove) && (move = maxMove)

    const movePercent = move / maxMove
    const newVal = positTurnVal(movePercent)

    if (newVal === lastVal.current) return
    updateInnerValue(newVal)
    lastVal.current = newVal
  }, 16)

  const handleSliderUp = () => {
    setMoving(false)
    window.removeEventListener('mousemove', handleSliderMove);
    window.removeEventListener('mouseup', handleSliderUp);
  };

  const getBtnRect = () => {
    const curRect = slider.current?.getBoundingClientRect()
    rect.left = curRect?.left || 0
    rect.top = curRect?.top || 0
    rect.width = curRect?.width || 0
    rect.height = curRect?.height || 0
    setRect({ ...rect })
  }

  const updateInnerValue = (val: number) => {
    if (!range) {
      setInnerValue(val)
    } else {
      if (Math.abs(val - (innerValue as number[])[0]) < Math.abs(val - (innerValue as number[])[1])) {
        (innerValue as number[])[0] = val
      } else {
        (innerValue as number[])[1] = val
      }
      setInnerValue([...innerValue as number[]])
    }
  }

  const handleSliderDown = (e: React.MouseEvent) => {
    if (!disabled) {
      e.persist();
      setMoving(true)
      getBtnRect()

      let movePercent = 0
      if (layout === 'horizontal') {
        movePercent = (e.clientX - rect.left) / rect.width
      } else {
        movePercent = (e.clientY - rect.top) / rect.height
      }
      const newVal = positTurnVal(movePercent)
      lastVal.current = newVal
      updateInnerValue(newVal)

      window.addEventListener('mousemove', handleSliderMove);
      window.addEventListener('mouseup', handleSliderUp);
    }
  }

  const singleSlider = (
    <>
      <div className="i-slider__bar">
        <div
          className="i-slider__bar-active"
          style={{
            left: layout === 'horizontal' ? `${(innerValue as number - min) / (max - min) * 100}%` : 0,
            top: layout === 'vertical' ? `${(innerValue as number - min) / (max - min) * 100}%` : 0,
          }}
        ></div>
      </div>
      <SliderBtn
        currentVal={innerValue as number}
        downSlider={downStatus}
        {...props}
      />
    </>
  )

  const rangeSlider = (
    <>
      <div className="i-slider__bar">
        <div
          className="i-slider__bar-active"
          style={{
            left: 0,
            top: 0,
            width: layout === 'horizontal' ? `${((innerValue as number[])[0] - min) / (max - min) * 100}%` : 6,
            height: layout === 'vertical' ? `${((innerValue as number[])[0] - min) / (max - min) * 100}%` : 6,
          }}
        ></div>
        <div
          className="i-slider__bar-active"
          style={{
            left: layout === 'horizontal' ? `${((innerValue as number[])[1] - min) / (max - min) * 100}%` : 0,
            top: layout === 'vertical' ? `${((innerValue as number[])[1] - min) / (max - min) * 100}%` : 0,
          }}
        ></div>
      </div>
      <SliderBtn
        currentVal={(innerValue as number[])[0]}
        downSlider={downStatus}
        {...props}
      />
      <SliderBtn
        currentVal={(innerValue as number[])[1]}
        downSlider={downStatus}
        {...props}
      />
    </>
  )

  return (
    <div
      className={classNames(
        'i-slider',
        layout === 'vertical' && `i-slider__layout-${layout}`,
        disabled && 'i-slider__disabled',
        className
      )}
      style={{ ...style }}
      ref={slider}
      onMouseDown={handleSliderDown}
    >
      {!range && !_.isArray(innerValue) && singleSlider}
      {range && _.isArray(innerValue) && rangeSlider}
    </div>
  );
};

Slider.displayName = 'Slider';

export default Slider;
