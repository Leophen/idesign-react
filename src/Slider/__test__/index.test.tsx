import { render } from '@testing-library/react';
import Slider from '../index';

describe('Slider 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<Slider />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('value', () => {
    const { container } = render(<Slider value={90} />);
    expect(container.querySelector('.i-slider__button')).toHaveStyle('left: calc(90% - 0px)');
  });

  it('defaultValue', () => {
    const { container } = render(<Slider defaultValue={90} />);
    expect(container.querySelector('.i-slider__button')).toHaveStyle('left: calc(90% - 0px)');
  });

  it('disabled', () => {
    const { container } = render(<Slider disabled />);
    expect(container.firstChild).toHaveClass('i-slider__disabled');
  });

  it('range', () => {
    const { asFragment } = render(<Slider range defaultValue={[30, 90]} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('vertical', () => {
    const { asFragment } = render(<Slider layout="vertical" />);
    expect(asFragment()).toMatchSnapshot();
  });
});

