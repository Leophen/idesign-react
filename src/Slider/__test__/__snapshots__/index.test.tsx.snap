// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-slider"
  >
    <div
      class="i-slider__bar"
    >
      <div
        class="i-slider__bar-active"
        style="left: 0%; top: 0px;"
      />
    </div>
    <div
      class="i-popup__reference"
    >
      <div
        class="i-slider__button"
        style="left: calc(0% - 0px);"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Slider 组件测试 range 1`] = `
<DocumentFragment>
  <div
    class="i-slider"
  >
    <div
      class="i-slider__bar"
    >
      <div
        class="i-slider__bar-active"
        style="left: 0px; top: 0px; width: 30%; height: 6px;"
      />
      <div
        class="i-slider__bar-active"
        style="left: 90%; top: 0px;"
      />
    </div>
    <div
      class="i-popup__reference"
    >
      <div
        class="i-slider__button"
        style="left: calc(30% - 0px);"
      />
    </div>
    <div
      class="i-popup__reference"
    >
      <div
        class="i-slider__button"
        style="left: calc(90% - 0px);"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Slider 组件测试 vertical 1`] = `
<DocumentFragment>
  <div
    class="i-slider i-slider__layout-vertical"
  >
    <div
      class="i-slider__bar"
    >
      <div
        class="i-slider__bar-active"
        style="left: 0px; top: 0%;"
      />
    </div>
    <div
      class="i-popup__reference"
    >
      <div
        class="i-slider__button"
        style="top: calc(0% - 0px);"
      />
    </div>
  </div>
</DocumentFragment>
`;
